### ISSUE TITLE
Deeply nested sqrt causes major slowdown

### ISSUE THREAD

**Issue by edemaine** [KaTeX/KaTeX#2788]
I just bumped into this year-old issue over on Gitlab:

https://gitlab.com/gitlab-org/gitlab-foss/-/issues/61410

Their 1.md example seems like something we should have an option to prevent. Apparently the SVGs from the sqrt cause a lot of rendering time. Maybe some kind of limit on nesting, at least when SVGs are involved?

**Comment by ronkok** [comment]
I doubt that SVGs are the main culprit. The Gitlab issue gives two examples and the [second one](https://gist.github.com/abdilahrf/959d75c5d39034ebc411844316a7ce4c#file-2-md) involves `\rule`, not `\sqrt`. `\rule` contains no SVG.

What both examples share is a call to the KaTeX internal function. `makeVList()`, which implements KaTeX vertical alignment. `makeVList` is remarkable. It does things that are difficult to do in HTML, but in so doing it creates a large number of visible HTML elements. The Gitlab examples each multiply that large number times a pathological number of `\sqrt` calls or `\rule` calls. Slowness ensues

Perhaps `makeVList` can be reimplemented in some more efficient way.  Or perhaps some limit on nesting anything that uses `makeVList`.

**Comment by ronkok** [comment]
... or maybe just a limit on nesting, full stop.

**Comment by ronkok** [comment]
Nope. The `\rule` example is not nested. It's just 9990 separate calls to `\rule`. A nesting limit would not prevent a DoS attack.

**Comment by ylemkimon** [comment]
Related: #2793

We should have some limit on nesting.

**Comment by ronkok** [comment]
Well, we could increment a counter every time `parseFunction()` is called. But that would not stop a determined attack. The [\rule example](https://gist.github.com/abdilahrf/959d75c5d39034ebc411844316a7ce4c#file-2-md) makes 9990 calls to `\rule` and each one has dimensions that are 80 digits long. 

That example has no nesting, but it still freezes execution.

**Comment by edemaine** [comment]
I thought the rule example was handled by `maxSize`? Or is it just as bad when the dimensions are small?

**Comment by ronkok** [comment]
@edemaine Thank you for reminding me about `maxsize`. I had forgotten. Yes, it will likely prevent the freeze resulting from the Gitlab `\rule` example. Of course, `maxsize` is currently set to `Infinity` which doesn't help much. And our [documentation](https://katex.org/docs/options.html) regarding `maxsize` does not mention that it was created to prevent a malicious attack. We should rewrite that section and make it more clear.

Then if we increment a counter in `parseFunction()` what should be the nesting limit? 1000? Set by a rendering option?

**Comment by edemaine** [comment]
Agreed. Perhaps the default for `maxSize` should change (can still be set to `Infinity` if desired), similar to our default limit for `maxExpand`. If it's set to something large like `1000em` does that make the performance OK?

I definitely think any other limits we impose should be settable by rendering options. Do we need to count function calls or nested function calls?  Does the `\sqrt` example also behave poorly when they are not nested? (How are you testing these?)

**Comment by ronkok** [comment]
Well, the `\sqrt` example makes 1,009,968 calls to `\sqrt`, in groups that are each nested 212 levels deep. I haven't tested the sensitivity to variations of nesting level vs pure volume, but I suspect that a malicious user could create a problem with either or both.
